File: go-services/internal/middleware/api_key.go
package middleware

import (
	"net/http"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
)

// APIKeyConfig holds API key middleware configuration
type APIKeyConfig struct {
	HeaderName   string
	APIKeys      map[string]string // api_key -> service_name
	SkipPaths    []string
}

// DefaultAPIKeyConfig returns default API key configuration
func DefaultAPIKeyConfig() APIKeyConfig {
	return APIKeyConfig{
		HeaderName: "X-API-Key",
		APIKeys:    make(map[string]string),
		SkipPaths: []string{
			"/health",
			"/ready",
			"/docs",
			"/swagger/",
			"/openapi.json",
		},
	}
}

// APIKeyAuth returns API key authentication middleware
func APIKeyAuth(cfg APIKeyConfig) gin.HandlerFunc {
	return func(c *gin.Context) {
		// Skip authentication for paths in skip list
		if shouldSkipPath(c.Request.URL.Path, cfg.SkipPaths) {
			c.Next()
			return
		}

		// Check if auth is disabled (no API keys configured)
		if len(cfg.APIKeys) == 0 {
			c.Next()
			return
		}

		apiKey := c.GetHeader(cfg.HeaderName)
		if apiKey == "" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error": "missing API key",
				"code":  "MISSING_API_KEY",
				"header": cfg.HeaderName,
			})
			return
		}

		// Validate API key
		serviceName, valid := cfg.APIKeys[apiKey]
		if !valid {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{
				"error": "invalid API key",
				"code":  "INVALID_API_KEY",
			})
			return
		}

		// Set service name in context
		c.Set("service", serviceName)
		c.Set("api_key", apiKey)

		c.Next()
	}
}

// APIKeyFromEnv creates APIKeyConfig from environment variables
// Format: API_KEYS=key1:service1,key2:service2
func APIKeyFromEnv() APIKeyConfig {
	cfg := DefaultAPIKeyConfig()
	cfg.HeaderName = getEnv("API_KEY_HEADER", "X-API-Key")

	apiKeysEnv := getEnv("API_KEYS", "")
	if apiKeysEnv != "" {
		pairs := strings.Split(apiKeysEnv, ",")
		for _, pair := range pairs {
			parts := strings.SplitN(pair, ":", 2)
			if len(parts) == 2 {
				cfg.APIKeys[parts[0]] = parts[1]
			}
		}
	}

	return cfg
}

// GetServiceName retrieves service name from context
func GetServiceName(c *gin.Context) string {
	if service, exists := c.Get("service"); exists {
		return service.(string)
	}
	return ""
}

// getEnv retrieves an environment variable or returns a default value.
func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

File: python-sentiment/app/middleware/api_key.py
"""API Key authentication middleware for Python sentiment service."""

import hmac

from fastapi import Request, HTTPException, status
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware

from app.config import get_settings


class APIKeyMiddleware(BaseHTTPMiddleware):
    """Middleware to validate API key for service-to-service communication."""

    def __init__(self, app, skip_paths: list = None):
        super().__init__(app)
        self.skip_paths = skip_paths or ["/health", "/ready", "/docs", "/redoc", "/openapi.json"]

    async def dispatch(self, request: Request, call_next):
        settings = get_settings()

        # Skip auth if disabled or no API key configured
        if not settings.enable_auth or not settings.api_key:
            return await call_next(request)

        # Skip auth for certain paths
        if request.url.path in self.skip_paths or request.url.path.startswith("/docs"):
            return await call_next(request)

        # Check API key
        api_key = request.headers.get("X-API-Key")
        if not api_key:
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={"error": "missing API key", "code": "MISSING_API_KEY"}
            )

        if not hmac.compare_digest(api_key, settings.api_key):
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={"error": "invalid API key", "code": "INVALID_API_KEY"}
            )

        return await call_next(request)
